pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -277,58 +326,97 @@ contract ManagedPool is ManagedPoolSettings {
-            request.amount = _upscale(request.amount, scalingFactorTokenOut);
+            request.amount = _upscale(request.amount, swapData.scalingFactorTokenOut);
 
             uint256 amountIn = WeightedMath._calcInGivenOut(
                 balanceTokenIn,
-                tokenInWeight,
+                swapData.tokenInWeight,
                 balanceTokenOut,
-                tokenOutWeight,
+                swapData.tokenOutWeight,
                 request.amount
             );
 
             // We round the amount in up (favoring a higher fee amount).
-            amountIn = amountIn.divUp(swapFeeComplement);
+            amountIn = amountIn.divUp(swapData.swapFeeComplement);
+
+            endingBalanceTokenIn = balanceTokenIn.add(amountIn);
+            endingBalanceTokenOut = balanceTokenOut.sub(request.amount);
 
             // amountIn tokens are entering the Pool, so we round up.
-            return _downscaleUp(amountIn, scalingFactorTokenIn);
+            amountCalculated = _downscaleUp(amountIn, swapData.scalingFactorTokenIn);
         }
+
+        // If circuit breakers are set, check the lower bound on the tokenIn, and the upper bound on the tokenOut.
+        _require(
+            !(CircuitBreakerLib.hasCircuitBreakerTripped(
+                _getCircuitBreakerState(request.tokenIn),
+                virtualSupply,
+                swapData.tokenInWeight,
+                endingBalanceTokenIn,
+                true // check lower bound
+            ) ||
+                CircuitBreakerLib.hasCircuitBreakerTripped(
+                    _getCircuitBreakerState(request.tokenOut),
+                    virtualSupply,
+                    swapData.tokenOutWeight,
+                    endingBalanceTokenOut,
+                    false // check upper bound
+                )),
+            Errors.CIRCUIT_BREAKER_TRIPPED
+        );
+    }
+
+    /**
+     * @dev Gather the information required to process a regular token swap, including circuit breaker bounds.
+     */
+    function _getSwapData(SwapRequest memory request, bytes32 poolState)
+        private
+        view
+        returns (SwapData memory tokenInfo)
+    {
+        bytes32 tokenInState = _getTokenState(request.tokenIn);
+        bytes32 tokenOutState = _getTokenState(request.tokenOut);
+
+        uint256 weightChangeProgress = ManagedPoolStorageLib.getGradualWeightChangeProgress(poolState);
+        tokenInfo.tokenInWeight = ManagedPoolTokenStorageLib.getTokenWeight(tokenInState, weightChangeProgress);
+        tokenInfo.tokenOutWeight = ManagedPoolTokenStorageLib.getTokenWeight(tokenOutState, weightChangeProgress);
+
+        tokenInfo.scalingFactorTokenIn = ManagedPoolTokenStorageLib.getTokenScalingFactor(tokenInState);
+        tokenInfo.scalingFactorTokenOut = ManagedPoolTokenStorageLib.getTokenScalingFactor(tokenOutState);
+
+        tokenInfo.swapFeeComplement = ManagedPoolStorageLib.getSwapFeePercentage(poolState).complement();
```

Comment:
It looks like we don't need this to be included in the struct but we can instead just load this inside `_onTokenSwap`. I'd be inclined to move it back so this struct only holds token related information (return variable name hints at this).

Reply:
It was another stack-too-deep thing. I can try moving some stuff back.